Type Around: add information about how to type before/after a widget (to be read by screen readers)

mmotyczynska opened the pull request:
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (widget): Screen readers should now read the keyboard shortcuts to type around a widget. Closes #11936.

---

### Additional information

Known issues:

*   after plugin change:isEnabled, information in the label is updated not immediately but after the first selection change
*   changing readonly mode does not affect the label

mlewand commented:
Based on what I know, though the solution have some edge cases - it solves most of the problems - so let's proceed with this proposal and extract the edge cases to a followup issue if it can't be fixed easily.

niegowski commented on packages/ckeditor5-widget/src/utils.ts:
```diff
@@ -225,11 +227,12 @@ export function setHighlightHandling(
-export function setLabel( element: Element, labelOrCreator: string | ( () => string ), writer: DowncastWriter ): void {
-	writer.setCustomProperty( 'widgetLabel', labelOrCreator, element );
+export function setLabel( element: Element, labelOrCreator: string | ( () => string ) ): void {
+	const widgetLabel = element.getCustomProperty( 'widgetLabel' ) as Array<string | Function>;
```
```suggestion
	const widgetLabel = element.getCustomProperty( 'widgetLabel' ) as Array<string | ( () => string )>;
```

niegowski commented on packages/ckeditor5-widget/src/utils.ts:
```diff
@@ -239,13 +242,15 @@ export function setLabel( element: Element, labelOrCreator: string | ( () => str
  */
 export function getLabel( element: Element ): string {
-	const labelCreator = element.getCustomProperty( 'widgetLabel' ) as string | ( () => string ) | undefined;
+	const widgetLabel = element.getCustomProperty( 'widgetLabel' ) as Array<string | Function>;
```
```suggestion
	const widgetLabel = element.getCustomProperty( 'widgetLabel' ) as Array<string | ( () => string )>;
```

niegowski commented on packages/ckeditor5-widget/src/utils.ts:
```diff
@@ -239,13 +242,15 @@ export function setLabel( element: Element, labelOrCreator: string | ( () => str
-	}
-
-	return typeof labelCreator == 'function' ? labelCreator() : labelCreator;
+	return widgetLabel.reduce( ( prev: string, current: string | Function ) => {
```
```suggestion
	return widgetLabel.reduce( ( prev: string, current: string | ( () => string ) ) => {
```

niegowski commented on packages/ckeditor5-widget/src/widgettypearound/widgettypearound.ts:
```diff
@@ -256,9 +256,19 @@ export default class WidgetTypeAround extends Plugin {
 			if ( isTypeAroundWidget( viewElement, data.item, schema ) ) {
 				injectUIIntoWidget( conversionApi.writer, buttonTitles, viewElement! );
+
+				const widgetLabel = viewElement.getCustomProperty( 'widgetLabel' ) as Array<string | Function>;
```
```suggestion
				const widgetLabel = viewElement.getCustomProperty( 'widgetLabel' ) as Array<string | ( () => string )>;
```

niegowski commented:
What about this comment? https://github.com/ckeditor/ckeditor5/issues/11936#issuecomment-1249216002

mmotyczynska commented:
> What about this comment? [#11936 (comment)](https://github.com/ckeditor/ckeditor5/issues/11936#issuecomment-1249216002)

I just added `aria-hidden` to these buttons, thanks for pointing that out :+1: 

Regarding the widget's `aria-describedby`, we added the description to the fake selection container instead because this is the one that is actually read by a screen reader when selecting a widget.

mlewand commented:
@niegowski are we waiting for something more or we're good to merge? 🙂

niegowski commented:
> @niegowski are we waiting for something more or we're good to merge? 🙂

I don't know the "bigger picture" and if this is a part of sth bigger or a standalone PR that could be merged right away.
